Update: SDK Demo Site Landing Page

jennsun opened the pull request:
https://app.clickup.com/t/y7h1rp

Updated the SDK Demo Site Landing Page to match that of the welcome page in the balsamiq mockup: https://balsamiq.cloud/sqswrp8/pmqxwaz/r2278. Hardcoded Default and Spectrum Reach Partner Configuration, as well as made a Custom Configuration for users to type in their own form configurations.

jennsun committed:
Update: SDK Demo Site Landing Page

jennsun committed:
Update: Fixed Login and Initialization Button

simanthastew commented on example_site/src/components/ConfigurationControls.css:
```diff
@@ -1,8 +1,22 @@
+.title-description {
+  display: grid;
+  place-items: start start;
```
If you only specify one value for `place-items`, it's also used as the second value, so a shorthand for this would be: 
`place-items: start;`

simanthastew commented on example_site/src/components/ConfigurationControls.css:
```diff
@@ -1,8 +1,22 @@
+  place-items: start start;
+  height: 600px;
+  grid-auto-flow: row;
+  background-color: white;
```
Since you set the background color as white in `App.css`, what happens if you don't specify it here?

simanthastew commented on example_site/src/components/ConfigurationControls.css:
```diff
@@ -1,8 +1,22 @@
   place-items: start start;
   height: 600px;
   grid-auto-flow: column;
+  background-color: white;
```
Same question about background color

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -24,6 +28,7 @@ export default function ConfigurationControls({ isOpen }) {
     openSnackbar,
   } = useAppContext();
 
+  
```
Let's get rid of this extra line

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -134,306 +150,348 @@ export default function ConfigurationControls({ isOpen }) {
+        
+        <button 
+          className="submit-button configuration-submit-button center-button">
+          {"See How It Works"}
```
You don't need the brackets or quotes here! Looks like we used to be using a ternary - in that scenario we have to wrap the expression in brackets so that JSX knows to evaluate it rather than just render the text, and then wrap strings in quotes. But if we're just rendering a simple string, neither are necessary.

`{waymarkInstance ? "Reset" : "See How It Works"}`

When JSX sees the above statement inside of the button element (or any element), and reads it as such:
1. If there is a `waymarkInstance`, render the text inside of the quotes after the `?`
2. Otherwise, render the text inside of the quotes after the `:`

simanthastew commented on example_site/src/components/Header.js:
```diff
@@ -0,0 +1,13 @@
+    )
+}
+
+export default Header;
```
Need a newline at the end of this file

simanthastew commented on example_site/src/components/Header.js:
```diff
@@ -0,0 +1,13 @@
+import React from 'react'
+ 
+function Header({title, subtitle}) {
```
Can you write a comment about what this component does? They're called docblocks, and we structure them like this:

```
/**
 * A description of what this component does and what it's used for
 * 
 * @param {string} title An explanation of what this prop is
 * @param {string} subtitle An explanation of what this prop is
 */
```
Let's go over a shortcut to get these started during our 1:1!

simanthastew commented on example_site/src/components/Header.js:
```diff
@@ -0,0 +1,13 @@
+        <div>
+            <h1>{title}</h1>
+            <h3>{subtitle}</h3>
+        </div>
```
It looks like the comps have the header taking up quite a bit more room than what we have here
![image](https://user-images.githubusercontent.com/22031072/123097689-5e2b9300-d3fe-11eb-9a77-2bd5e05e7cde.png)
![image](https://user-images.githubusercontent.com/22031072/123097746-6c79af00-d3fe-11eb-8e0b-3b616bfab765.png)

So you might want to use CSS instead of just header tags to achieve this. You can create a new file, `Header.css` and import the styles just like we do in other components, i.e. `import "./Header.css` and then apply any styles you define.

simanthastew replied:
Also, look up JSX fragments as an alternative to wrapping this component in a `<div>`!

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -134,306 +150,348 @@ export default function ConfigurationControls({ isOpen }) {
+            </div>
           </div>
-        </div>
+        </div>   
```
That darker green swatch to the right of `</div>` here means there is some whitespace, you'll want to get rid of that!

simanthastew commented on example_site/src/components/ConfigurationControls.css:
```diff
@@ -1,8 +1,22 @@
 .configuration-controls-form {
   display: grid;
   place-items: start start;
   height: 600px;
```
This was old, but I'm seeing the bottom of the custom form get cut off. I'm not sure why we were using this `600px` height, but let's get rid of it here and in the `title-description` class (and anywhere else) and let this page be full height

simanthastew commented on example_site/src/components/ConfigurationControls.css:
```diff
@@ -20,6 +34,33 @@ button.configuration-submit-button {
+.center-title {
+  margin: auto;
+  text-align: center;
+  width: 40%;
```
Why 40%?

simanthastew commented on example_site/src/components/ConfigurationControls.css:
```diff
@@ -20,6 +34,33 @@ button.configuration-submit-button {
+.center-form {
+  margin: auto;
+  text-align: left;
+  width: 80%;
```
Why 80%?

jennsun replied:
In terms of the form spacing, the following are with and without the width set to 80%
![image](https://user-images.githubusercontent.com/26675662/123128402-1d8d4300-d419-11eb-9433-4536e08e579b.png)
![image](https://user-images.githubusercontent.com/26675662/123128461-2a119b80-d419-11eb-899b-bf54bf981764.png)

I will keep the width at 80% for now for this purpose but may be subject to change in the future

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -50,7 +55,15 @@ export default function ConfigurationControls({ isOpen }) {
+    if (formData.partner === 'default')
+      configuration = defaultConfiguration;
+    else if (formData.partner === 'spectrum')
+      configuration = spectrumConfiguration;
```
This works really well when we only have three options, but what if we were to add more options in the future? Say we had 10 additional partners in the dropdown, would this be easy to maintain? Let's brainstorm other ways we could accomplish this that are a little more flexible during our chat.

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -50,7 +55,15 @@ export default function ConfigurationControls({ isOpen }) {
+      configuration = spectrumConfiguration;
+
     const {
+      partner,
```
This `partner` variable is not required in the configuration - it's just a value that we're using to decide which configuration to use. We can get rid of it here, and in the configuration constants.

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -117,7 +133,7 @@ export default function ConfigurationControls({ isOpen }) {
         backgroundColor: editorBackgroundColor,
       },
-      environment,
+      environment: environment,
```
When using a variable as a value in an object with the same name as its key, you actually don't need to explicitly state the value, you can leave it like we had it. This is called [object property shorthand](https://alligator.io/js/object-property-shorthand-es6/) and it's pretty cool!

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -134,306 +150,348 @@ export default function ConfigurationControls({ isOpen }) {
-      <div>
-        <h2>Configuration</h2>
+    <form onSubmit={handleSubmit(onSubmit)} className={titlePanel}>
+      <div className='center-title'>
+        <div className='Header'>
```
Do you need both divs here? Each element has its own properties and rules and will affect the layout of your page, so it's good practice to only include elements that you need to achieve your layout.

jennsun replied:
No, the div containing the className='Header' was not needed!

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -134,306 +150,348 @@ export default function ConfigurationControls({ isOpen }) {
-          <option value="demo">Demo</option>
-          <option value="prod">Production</option>
-          <option value="local">Local</option>
+          <option selected value="default">Default Partner</option>
+          <option value="spectrum">Spectrum Reach</option>
+          <option value="custom">Custom</option>
```
We've been bad developers (me ryan and scott, not you)! We should define these partners, and the environments from the environment dropdown as constants like we do for the configurations.

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -84,7 +97,10 @@ export default function ConfigurationControls({ isOpen }) {
 
     // Correct orientation for editor
-    const finalOrientation = orientation ? "right" : "left";
+    let finalOrientation;
+    if (orientation===true) finalOrientation = 'right';
+    else if (!orientation) finalOrientation = 'left';
+    else finalOrientation = orientation; 
```
This is getting tricky isn't it! A few things to pay attention to here:
1. It's tempting to leave out the closures (brackets) for these if statements, but in my opinion it does make this code block a little less readable, and makes it easier for errors to slip through.
2. Since your final else statement assigns `finalOrientation` to `orientation`, you could use that as the initialization value on line 100 instead of using the `else` statement.
3. I think that the real trickiness is coming from having our orientation value be in two different formats (true vs. false/right vs.left) so let's figure out a cleaner way to get this value!

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -134,306 +150,348 @@ export default function ConfigurationControls({ isOpen }) {
-    panel: true,
-    open: isOpen,
-    closed: !isOpen,
+    'modal-fade': true,
```
Since this is no longer just being used for modals, let's broaden up the classname, like `fade-in-out` or something like that.

simanthastew reviewed (changes requested):
Really good stuff here! Some small stuff, but a few areas I think we could refactor and clean up the logic a bit. Excited to go over all of this with you!

jennsun committed:
PR Review
Reduced redundancy for place-items, removed background color in Configuration Controls because it was already set in App.css, addded docblocks to new functions/Header component, fixed height so form does not require scrolling, changed title/subtitle style to better match in the balsamiq mockup, implemented mapping method to determine form configurations, created constants for the chosen configurations, cleaned up issues relating to naming/dividng code into functions

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -96,11 +115,7 @@ export default function ConfigurationControls({ isOpen }) {
-    if (orientation===true) finalOrientation = 'right';
-    else if (!orientation) finalOrientation = 'left';
-    else finalOrientation = orientation; 
+    const finalOrientation = getOrientation(orientation);
```
Since you're only using the variable `finalOrientation` once, you could just use the `getOrientation(orientation)` call in the `waymarkOptions` object itself, i.e., `orientation: getOrientation(orientation)`. 

^^ This is _very_ nitpicky, just highlighting your different options!

simanthastew commented on example_site/src/components/Header.css:
```diff
@@ -0,0 +1,7 @@
+
+.subtitle {
+    font-size: 25px;
+}
```
Newline!

simanthastew commented on example_site/src/components/constants.js:
```diff
@@ -46,4 +50,9 @@ export const spectrumConfiguration = {
+export const partnerConfigurations = {
+    [partnerPresets.default]: defaultConfiguration,
+    [partnerPresets.spectrum]: spectrumConfiguration,
+};
```
newline!

simanthastew commented on example_site/src/components/Header.js:
```diff
@@ -1,13 +1,20 @@
+import "./Header.css"
+
+/**
+ * Creates a Header reusable component
```
What about something like, "Renders a title and subtitle, intended to be used as a header"?

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -54,16 +53,36 @@ export default function ConfigurationControls({ isOpen }) {
-  const onSubmit = async (formData) => {
-    let configuration = formData;
+  /**
+   * Returns proper configuration settings for SDK Demo
```
I would put a little something about the partner in here - it will shed a little light about what the "proper" configuration settings are.

simanthastew commented on example_site/src/components/ConfigurationControls.js:
```diff
@@ -54,16 +53,36 @@ export default function ConfigurationControls({ isOpen }) {
+    if (orientation === 'left' || orientation === 'right') {
+      return orientation;
+    }
+    // Custom configuration results in value of false when user
+    // desires left orientation and true when user desires
+    // right orientation due to checkbox implementation
```
Amazing comment :)

simanthastew reviewed (changes requested):
This is looking great - just a few very very small things, feel free to merge when you're done!

One note about your most recent commit message - we typically like to structure the additional content like this:

```
Main commit message

- Additional detail
- Another detail
- One more detail for good measure
```

No need to redo it, just keep it in mind next time!

jennsun committed:
PR Review 2
- Directly used function call to get proper orientation value instead of setting it to a variable
- Added new lines at end of files
- Included more in-depth comments
